types/helpers.d.ts

```diff
@@ -1,69 +1,100 @@
-type Computed = () => any;
-type MutationMethod = (...args: any[]) => void;
-type ActionMethod = (...args: any[]) => Promise<any>;
-type CustomVue = Vue & Dictionary<any>
-
-interface Mapper<R> {
-  (map: string[]): Dictionary<R>;
-  (map: Dictionary<string>): Dictionary<R>;
+/**
+ * Utility types to declare helper types
+ */
+type Computed<R> = () => R;
+type Method<R> = (...args: any[]) => R;
+type MutationMethod<P> = (payload: P) => void;
+type ActionMethod<P> = (payload: P) => Promise<any>;
+type CustomVue = Vue & Record<string, any>;
+
+interface BaseType { [key: string]: any }
+
+interface BaseStateMap<State, Getters> {
+  [key: string]: (this: CustomVue, state: State, getters: Getters) => any;
+}
+
+interface BaseMethodMap<F> {
+  [key: string]: (this: CustomVue, fn: F, ...args: any[]) => any;
+}
+
+/**
+ * mapGetters
+ */
+interface MapGetters<Getters> {
+  <G extends Getters = Getters, Key extends keyof G = keyof G>(map: Key[]): { [K in Key]: Computed<G[K]> };
+  <G extends Getters = Getters, Map extends Record<string, keyof G> = Record<string, keyof G>>(map: Map): { [K in keyof Map]: Computed<G[Map[K]]> };
+}
+
+interface RootMapGetters<Getters> extends MapGetters<Getters> {
+  <G extends Getters = Getters, Key extends keyof G = keyof G>(namespace: string, map: Key[]): { [K in Key]: Computed<G[K]> };
+  <G extends Getters = Getters, Map extends Record<string, keyof G> = Record<string, keyof G>>(namespace: string, map: Map): { [K in keyof Map]: Computed<G[Map[K]]> };
+}
+
+/**
+ * mapState
+ */
+interface MapState<State, Getters> {
+  <S extends State = State, Key extends keyof S = keyof S>(map: Key[]): { [K in Key]: Computed<S[K]> };
+  <S extends State = State, Map extends Record<string, keyof S> = Record<string, keyof S>>(map: Map): { [K in keyof Map]: Computed<S[Map[K]]> };
+  <S extends State = State, G extends Getters = Getters, Map extends BaseStateMap<S, G> = BaseStateMap<S, G>>(map: Map): { [K in keyof Map]: Computed<any> };
 }
 
-interface MapperWithNamespace<R> {
-  (namespace: string, map: string[]): Dictionary<R>;
-  (namespace: string, map: Dictionary<string>): Dictionary<R>;
+interface RootMapState<State, Getters> extends MapState<State, Getters> {
+  <S extends State = State, Key extends keyof S = keyof S>(namespace: string, map: Key[]): { [K in Key]: Computed<S[K]> };
+  <S extends State = State, Map extends Record<string, keyof S> = Record<string, keyof S>>(namespace: string, map: Map): { [K in keyof Map]: Computed<S[Map[K]]> };
+  <S extends State = State, G extends Getters = Getters, Map extends BaseStateMap<S, G> = BaseStateMap<S, G>>(namespace: string, map: Map): { [K in keyof Map]: Computed<any> };
 }
 
-interface FunctionMapper<F, R> {
-  (map: Dictionary<(this: CustomVue, fn: F, ...args: any[]) => any>): Dictionary<R>;
+/**
+ * mapMutations
+ */
+interface MapMutations<Mutations> {
+  <M extends Mutations = Mutations, Key extends keyof M = keyof M>(map: Key[]): { [K in Key]: MutationMethod<M[K]> };
```

Comment:
Usually generic default is for interface / type alias. Mapper is not exported nor augmented by users, so I think these defaults aren't required.

Reply:
Actually, it throws error if `Mutations` is `BaseType` since `M` will be inferred as `{}` in that case. I have no idea how we fix it without default generic type 😞 

But we may no longer need it if we create typed root helpers with `createNamespacedHelpers` in any cases.

Reply:
Let's try typed root helpers! It looks more promising since it can return more precise types.